gds_write/def2stream.py

```diff
@@ -0,0 +1,208 @@
+parser.add_argument("-i", "--input-def",      required=True,  help="Path to def file of the implemented design after detailed routing step")
+parser.add_argument("-t", "--tech-file",      required=True,  help="Path to KLayout technology file, a mapping from LEF/DEF to GDS.")
+parser.add_argument("-o", "--out",            required=True,  help="Path to output GDS file")
+parser.add_argument("-g", "--additional-gds", required=False, help="Paths to additional input GDS files", default="", nargs="*")
+parser.add_argument("-f", "--fill-config",    required=False, help="Path to JSON rule file for metal fill during chip finishing", default="")
+parser.add_argument("-s", "--seal",           required=False, help="Path to seal GDS/OAS file", default="")
+parser.add_argument("-l", "--layer-map",      required=False, help="Path to layer map file", default="")
+
+args = parser.parse_args()
+
+errors = 0
+
+# Expand layers in json
+def expand_cfg_layers(cfg):
+  layers = cfg['layers']
+  expand = [layer for layer in layers if 'layers' in layers[layer]]
+  for layer in expand:
+    for i, (name, num) in enumerate(zip(layers[layer]['names'],
+                                        layers[layer]['layers'])):
+      new_layer = copy.deepcopy(layers[layer])
+      del new_layer['names']
+      new_layer['name'] = name
+      del new_layer['layers']
+      new_layer['layer'] = num
+      layers[name] = new_layer
+    del layers[layer]
+
+def read_cfg():
+  print('INFO: Reading config file: ' + args.fill_config)
+  with open(args.fill_config, 'r') as f:
+    cfg = json.load(f)
+
+  expand_cfg_layers(cfg)
+  cfg = cfg['layers'] # ignore the rest
+
+  # Map gds layers & datatype to KLayout indices
+  # These are arrays for the different mask numbers
+  for layer, vals in cfg.items():
+    layer = vals['layer']
+    for key in ('opc', 'non-opc'):
+      if key not in vals:
+        continue
+      data = vals[key]
+      if isinstance(data['datatype'], int):
+        data['datatype'] = [data['datatype']] # convert to array
+      data['klayout'] = [main_layout.find_layer(layer, datatype)
+                         for datatype in data['datatype']]
+
+  return cfg
+
+#match a line like:
+# - LAYER M2 + MASK 2 + OPC RECT ( 3000 3000 ) ( 5000 5000 ) ;
+rect_pat = re.compile(r'''
+  \s*\-\ LAYER\ (?P<layer>\S+)  # The layer name
+  (?:                           # Non-capturing group
+  \s+\+\ MASK\ (?P<mask>\d+)    # Mask, None if absent
+  )?
+  (?P<opc>                      # OPC, None if absent
+  \s+\+\ OPC
+  )?
+  \s+RECT\
+   \(\ (?P<xlo>\d+)\ (?P<ylo>\d+)\ \)\   # rect lower-left pt
+  \(\ (?P<xhi>\d+)\ (?P<yhi>\d+)\ \)\ ; # rect upper-right pt
+  ''',
+                      re.VERBOSE)
+
+def read_fills(top):
+  if args.fill_config == '':
+    print('WARNING: no fill config file specified')
+    return
+  # KLayout doesn't support FILL in DEF so we have to side load them :(
```

Comment:
is there an issue on the klayout repo already opened for that?

Reply:
Looks like this issue was solved in https://github.com/KLayout/klayout/issues/673. I modified the script and bazel flow to perform density fill after detailed routing and then write final DEF file which is read during GDS write